src/lib/core/cbus.c

```diff
@@ -177,6 +187,145 @@ cpipe_destroy(struct cpipe *pipe)
 	TRASH(pipe);
 }
 
+struct lcpipe *
+lcpipe_new(const char *consumer)
+{
+	struct lcpipe *pipe = xmalloc(sizeof(*pipe));
+
+	stailq_create(&pipe->input);
+
+	pipe->n_input = 0;
+	pipe->max_input = INT_MAX;
+
+	acquire_consumer(&pipe->endpoint, consumer);
+
+	return pipe;
+}
+
+/**
+ * Flush all staged messages into consumer output and wake up it after.
+ */
+static void
+lcpipe_flush(struct lcpipe *pipe)
+{
+	struct cbus_endpoint *endpoint = pipe->endpoint;
+	if (pipe->n_input == 0)
+		return;
+
+	/* Trigger task processing when the queue becomes non-empty. */
+	bool output_was_empty;
+
+	/*
+	 * We need to set a thread cancellation guard, because
+	 * another thread may cancel the current thread
+	 * (write() is a cancellation point in ev_async_send)
+	 * and the activation of the ev_async watcher
+	 * through ev_async_send will fail.
+	 */
+
+	int old_cancel_state;
+	tt_pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &old_cancel_state);
+
+	tt_pthread_mutex_lock(&endpoint->mutex);
+	output_was_empty = stailq_empty(&endpoint->output);
+	/** Flush input */
+	stailq_concat(&endpoint->output, &pipe->input);
+	tt_pthread_mutex_unlock(&endpoint->mutex);
+
+	pipe->n_input = 0;
+	if (output_was_empty) {
+		/* Count statistics */
+		rmean_collect(cbus.stats, CBUS_STAT_EVENTS, 1);
+
+		ev_async_send(pipe->endpoint->consumer, &pipe->endpoint->async);
+	}
+
+	tt_pthread_setcancelstate(old_cancel_state, NULL);
+}
+
+void
+lcpipe_flush_input(struct lcpipe *pipe)
+{
+	/** Flush may be called with no input. */
+	if (pipe->n_input > 0) {
+		lcpipe_flush(pipe);
+	}
+}
+
+/**
+ * Insert message into pipe input queue.
+ */
+static inline void
+lcpipe_insert_input(struct lcpipe *pipe, struct cmsg *msg)
+{
+	assert(msg->hop->pipe == NULL);
+	stailq_add_tail_entry(&pipe->input, msg, fifo);
+	pipe->n_input++;
+	if (pipe->n_input >= pipe->max_input) {
+		lcpipe_flush(pipe);
+	}
+}
```

Comment:
This code is identical to `cpipe_push_input` with the only difference that you call here `lcpipe_flush` and the former calls `ev_invoke(pipe->producer, &pipe->flush_input, EV_CUSTOM);`. But if you look at what `ev_invoke` does, you will find out, that it calls `cpipe_flush_cb` here which is the same as your `lcpipe_flush`.

Reply:
There is a fundamental difference between cpipe_push_input and lcpipe_insert_input: the second can be called in any context, but the first can only be called in cord.
There are some differences in the code:
- cpipe_push_input checks for the presence of a loop (we need cord for this) in the producer side: ```assert(loop() == pipe->producer);```
- lcpipe_insert_input calls lcpipe_flush for which only lcpipe itself is needed, it cannot call cpipe_flush_cb because it requires ev_loop and ev_async which are not available outside the cord context

Reply:
> cpipe_push_input checks for the presence of a loop (we need cord for this) in the producer side: assert(loop() == pipe->producer);

And what is the problem with that? The idea is not to make cpipe and lcpipe use exactly the same methods but reuse the main part of the code. This assertion you can simply leave at the cpipe level:
```C
void
lcpipe_push(struct lcpipe *pipe, struct cmsg *msg)
{
    assert(msg->hop->pipe == NULL);
    stailq_add_tail_entry(&pipe->base.input, msg, fifo);
    pipe->base.n_input++;
    if (pipe->base.n_input >= pipe->base.max_input) {
        lcpipe_flush_input(pipe);
    }
}

static inline void
cpipe_push_input(struct cpipe *pipe, struct cmsg *msg)
{
    assert(loop() == pipe->producer);
    lcpipe_push(&pipe->base, msg);
}
```

As for the triggers - I noted in another commit, that these on-flush triggers are unused. Delete them as a preparatory commit, and you will be able to reuse the code like I did above.

> it cannot call cpipe_flush_cb because it requires ev_loop and ev_async which are not available outside the cord context

Where does it require them? If you mean `loop()`, then do like I've shown in the code above - keep it at the cpipe level. If you mean `ev_invoke()`, then I explained it in my first comment in this thread, that it isn't needed even in the cpipe. Fix it as another preparatory commit, and you will be able to reuse the code effortlessly.